src/peer_manager.rs

```diff
@@ -981,14 +982,33 @@ impl PeerManager {
             None
         };
 
+        let mut demote_peers = Vec::new();
         let expired_peers = self.peers
             .values()
-            .filter(|peer| peer.is_expired())
+            .filter(|peer| {
+                if peer.is_expired() {
+                    match peer.state {
+                        PeerState::Routing(RoutingConnection::JoiningNode(_)) |
+                        PeerState::Routing(RoutingConnection::Proxy(_)) => {
+                            demote_peers.push(*peer.pub_id());
+                            return false;
+                        }
+                        _ => return true,
+                    }
+                }
+                false
+            })
             .map(Peer::pub_id)
             .cloned()
             .chain(remove_candidate)
             .collect_vec();
 
+        for id in &demote_peers {
```

Comment:
The term demotion does not sound right - it's neither a demotion or a promotion in the sense that it does not get any more special status (it's already a routing connection). Rather it's slightly a promotion if anything that now it's allowed to act as a tunnel to its (previously) proxy/joining-node. So maybe should be named something different (e.g. `update_to_routing_direct` or something - see if you can come up with a better name).